Subject: [PATCH] forbiddenFunctions: bypass noexcept on Bionic libc

JDK 26 added forbiddenFunctions.hpp which redeclares standard libc
functions like strerror, malloc, wcsdup with explicit noexcept.
Bionic (Android) declares these without noexcept, causing
"exception specification in declaration does not match" errors.

Introduce a BIONIC_NOEXCEPT macro that resolves to noexcept on
non-Android targets and to nothing on Android, and apply it to all
forbid declarations.

--- a/src/hotspot/share/utilities/forbiddenFunctions.hpp
+++ b/src/hotspot/share/utilities/forbiddenFunctions.hpp
@@ -44,40 +44,49 @@
 #include "forbiddenFunctions_posix.hpp"
 #endif
 
+
+// Bionic libc does not declare standard C functions with noexcept,
+// so on Android the noexcept on these forbids causes a redeclaration mismatch.
+#if defined(__BIONIC__) || defined(__ANDROID__)
+#define BIONIC_NOEXCEPT
+#else
+#define BIONIC_NOEXCEPT noexcept
+#endif
+
 // Forbid the use of various C library functions.  Some of these have os::
 // replacements that should be used instead.  Others are considered obsolete
 // or have security concerns, either with preferred alternatives, or to be
 // avoided entirely.
 
-FORBID_IMPORTED_NORETURN_C_FUNCTION(void exit(int), noexcept, "use os::exit")
-FORBID_IMPORTED_NORETURN_C_FUNCTION(void _Exit(int), noexcept, "use os::exit")
+FORBID_IMPORTED_NORETURN_C_FUNCTION(void exit(int), BIONIC_NOEXCEPT, "use os::exit")
+FORBID_IMPORTED_NORETURN_C_FUNCTION(void _Exit(int), BIONIC_NOEXCEPT, "use os::exit")
 
 // Windows puts _exit in <stdlib.h>, POSIX in <unistd.h>.
 FORBID_IMPORTED_NORETURN_C_FUNCTION(void _exit(int), /* not noexcept */, "use os::exit")
 
-FORBID_IMPORTED_C_FUNCTION(char* strerror(int), noexcept, "use os::strerror");
-FORBID_IMPORTED_C_FUNCTION(char* strtok(char*, const char*), noexcept, "use strtok_r");
+FORBID_IMPORTED_C_FUNCTION(char* strerror(int), BIONIC_NOEXCEPT, "use os::strerror");
+FORBID_IMPORTED_C_FUNCTION(char* strtok(char*, const char*), BIONIC_NOEXCEPT, "use strtok_r");
 
 // AIX declarations for sprintf and snprintf are not noexcept, which is
 // inconsistent with most other system header declarations, including being
 // inconsistent with vsprintf and fsnprintf.
-FORBID_C_FUNCTION(int sprintf(char*, const char*, ...), NOT_AIX(noexcept), "use os::snprintf");
-FORBID_C_FUNCTION(int snprintf(char*, size_t, const char*, ...), NOT_AIX(noexcept), "use os::snprintf");
+FORBID_C_FUNCTION(int sprintf(char*, const char*, ...), NOT_AIX(BIONIC_NOEXCEPT), "use os::snprintf");
+FORBID_C_FUNCTION(int snprintf(char*, size_t, const char*, ...), NOT_AIX(BIONIC_NOEXCEPT), "use os::snprintf");
 
 PRAGMA_DIAG_PUSH
 FORBIDDEN_FUNCTION_IGNORE_CLANG_FORTIFY_WARNING
-FORBID_C_FUNCTION(int vsprintf(char*, const char*, va_list), noexcept, "use os::vsnprintf");
-FORBID_C_FUNCTION(int vsnprintf(char*, size_t, const char*, va_list), noexcept, "use os::vsnprintf");
+FORBID_C_FUNCTION(int vsprintf(char*, const char*, va_list), BIONIC_NOEXCEPT, "use os::vsnprintf");
+FORBID_C_FUNCTION(int vsnprintf(char*, size_t, const char*, va_list), BIONIC_NOEXCEPT, "use os::vsnprintf");
 PRAGMA_DIAG_POP
 
 // All of the following functions return raw C-heap pointers (sometimes as an
 // option, e.g. realpath or getwd) or, in case of free(), take raw C-heap
 // pointers.  We generally want allocation to be done through NMT.
-FORBID_IMPORTED_C_FUNCTION(void* malloc(size_t size), noexcept, "use os::malloc");
-FORBID_IMPORTED_C_FUNCTION(void free(void *ptr), noexcept, "use os::free");
-FORBID_IMPORTED_C_FUNCTION(void* calloc(size_t nmemb, size_t size), noexcept, "use os::malloc and zero out manually");
-FORBID_IMPORTED_C_FUNCTION(void* realloc(void *ptr, size_t size), noexcept, "use os::realloc");
-FORBID_IMPORTED_C_FUNCTION(char* strdup(const char *s), noexcept, "use os::strdup");
-FORBID_IMPORTED_C_FUNCTION(wchar_t* wcsdup(const wchar_t *s), noexcept, "don't use");
+FORBID_IMPORTED_C_FUNCTION(void* malloc(size_t size), BIONIC_NOEXCEPT, "use os::malloc");
+FORBID_IMPORTED_C_FUNCTION(void free(void *ptr), BIONIC_NOEXCEPT, "use os::free");
+FORBID_IMPORTED_C_FUNCTION(void* calloc(size_t nmemb, size_t size), BIONIC_NOEXCEPT, "use os::malloc and zero out manually");
+FORBID_IMPORTED_C_FUNCTION(void* realloc(void *ptr, size_t size), BIONIC_NOEXCEPT, "use os::realloc");
+FORBID_IMPORTED_C_FUNCTION(char* strdup(const char *s), BIONIC_NOEXCEPT, "use os::strdup");
+FORBID_IMPORTED_C_FUNCTION(wchar_t* wcsdup(const wchar_t *s), BIONIC_NOEXCEPT, "don't use");
 
 #endif // SHARE_UTILITIES_FORBIDDENFUNCTIONS_HPP
--- a/src/hotspot/os/posix/forbiddenFunctions_posix.hpp
+++ b/src/hotspot/os/posix/forbiddenFunctions_posix.hpp
@@ -37,26 +37,26 @@
 #endif
 
 // If needed, add os::strndup and use that instead.
-FORBID_C_FUNCTION(char* strndup(const char*, size_t), noexcept, "don't use");
+FORBID_C_FUNCTION(char* strndup(const char*, size_t), BIONIC_NOEXCEPT, "don't use");
 
 // These are unimplementable for Windows, and they aren't useful for a
 // POSIX implementation of NMT either.
 // https://stackoverflow.com/questions/62962839/stdaligned-alloc-missing-from-visual-studio-2019
-FORBID_C_FUNCTION(int posix_memalign(void**, size_t, size_t), noexcept, "don't use");
-FORBID_C_FUNCTION(void* aligned_alloc(size_t, size_t), noexcept, "don't use");
+FORBID_C_FUNCTION(int posix_memalign(void**, size_t, size_t), BIONIC_NOEXCEPT, "don't use");
+FORBID_C_FUNCTION(void* aligned_alloc(size_t, size_t), BIONIC_NOEXCEPT, "don't use");
 
 // realpath with a null second argument mallocs a string for the result.
 // With a non-null second argument, there is a risk of buffer overrun.
 PRAGMA_DIAG_PUSH
 FORBIDDEN_FUNCTION_IGNORE_CLANG_FORTIFY_WARNING
-FORBID_C_FUNCTION(char* realpath(const char*, char*), noexcept, "use os::realpath");
+FORBID_C_FUNCTION(char* realpath(const char*, char*), BIONIC_NOEXCEPT, "use os::realpath");
 PRAGMA_DIAG_POP
 
 // Returns a malloc'ed string.
-FORBID_C_FUNCTION(char* get_current_dir_name(), noexcept, "use os::get_current_directory");
+FORBID_C_FUNCTION(char* get_current_dir_name(), BIONIC_NOEXCEPT, "use os::get_current_directory");
 
 // Problematic API that should never be used.
-FORBID_C_FUNCTION(char* getwd(char*), noexcept, "use os::get_current_directory");
+FORBID_C_FUNCTION(char* getwd(char*), BIONIC_NOEXCEPT, "use os::get_current_directory");
 
 // BSD utility that is subtly different from realloc.
 FORBID_C_FUNCTION(void* reallocf(void*, size_t), /* not noexcept */, "use os::realloc");
